File: src/io/github/bagu/dough/parser/Parser.java
package io.github.bagu.dough.parser;

import io.github.bagu.dough.ast.*;
import vm.ast.*;
import io.github.bagu.dough.source.SourceRange;
import io.github.bagu.dough.token.Token;
import io.github.bagu.dough.token.TokenKind;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class Parser {
    private static final Map<TokenKind, BinaryOp> BINARY_OPS;
    private static final Map<TokenKind, UnaryOp> UNARY_OPS;
    private static final Set<TokenKind> LITERALS;

    static {
        BINARY_OPS = Map.ofEntries(
                Map.entry(TokenKind.STAR, BinaryOp.MUL),
                Map.entry(TokenKind.SLASH, BinaryOp.DIV),
                Map.entry(TokenKind.PLUS, BinaryOp.ADD),
                Map.entry(TokenKind.MINUS, BinaryOp.SUB),

                Map.entry(TokenKind.GT, BinaryOp.GT),
                Map.entry(TokenKind.LT, BinaryOp.LT),
                Map.entry(TokenKind.GE, BinaryOp.GE),
                Map.entry(TokenKind.LE, BinaryOp.LE),

                Map.entry(TokenKind.EQ, BinaryOp.EQ),
                Map.entry(TokenKind.NEQ, BinaryOp.NEQ),

                Map.entry(TokenKind.AND, BinaryOp.AND),
                Map.entry(TokenKind.OR, BinaryOp.OR)
        );

        UNARY_OPS = Map.ofEntries(
                Map.entry(TokenKind.NOT, UnaryOp.NOT),
                Map.entry(TokenKind.MINUS, UnaryOp.NEG)
        );

        LITERALS = Set.of(
                TokenKind.STR_LIT,
                TokenKind.INT_LIT,
                TokenKind.FLOAT_LIT,
                TokenKind.TRUE,
                TokenKind.FALSE
        );
    }

    private final List<Token> tokens;

    private int current = 0;

    public Parser(List<Token> tokens) {
        this.tokens = List.copyOf(tokens);

        if (tokens.isEmpty())
            throw new IllegalArgumentException("Expected terminating EOF token");

        Token last = tokens.getLast();
        if (last.kind() != TokenKind.EOF)
            throw new IllegalArgumentException("Expected terminating EOF token");

        for (Token token : tokens) {
            if (token == last) break;

            if (token.kind() == TokenKind.EOF)
                throw new IllegalArgumentException("Unexpected EOF token at index " + tokens.indexOf(token));
        }
    }

    public Program parse() {
        current = 0;
        List<Decl> decls = new ArrayList<>();

        while (!atEnd()) {
            decls.add(parseDecl());
        }

        return new Program(decls, range(0));
    }

    private ParseException exception(String message) {
        return new ParseException(message, peek(0).range());
    }

    private SourceRange range(int start) {
        return new SourceRange(start, previous().range().end());
    }

    private int start() {
        return peek(0).range().start();
    }

    private boolean atEnd() {
        return tokens.get(current).kind() == TokenKind.EOF;
    }

    private Token advance() {
        return atEnd()
                ? tokens.getLast()
                : tokens.get(current++);
    }

    private Token peek(int offset) {
        int index = current + offset;
        return (index >= tokens.size())
                ? tokens.getLast()
                : tokens.get(index);
    }

    private Token previous() {
        int index = current - 1;
        if (index < 0) return tokens.getFirst();
        if (index >= tokens.size()) return tokens.getLast();
        return tokens.get(index);
    }

    private boolean check(TokenKind expected) {
        return peek(0).kind() == expected;
    }

    private boolean match(TokenKind expected) {
        boolean matches = check(expected);
        if (matches) advance();
        return matches;
    }

    private void skip(int count) {
        for (int i = 0; i < count; i++) {
            advance();
        }
    }

    private Token expect(TokenKind kind, String message) {
        if (check(kind)) return advance();
        throw exception(message);
    }

    private TypeRef expectTypeRef() {
        Token token = peek(0);
        return switch (token.kind()) {
            case STR, INT, FLOAT, BOOL, VOID, IDENT -> new TypeRef(advance().lexeme(), token.range());
            default -> throw exception("Expected type");
        };
    }

    private Decl parseDecl() {
        return switch (peek(0).kind()) {
            case VAR -> parseVarDecl();
            case FUNC -> parseFuncDef();

            default -> throw exception("Unexpected token \"" + peek(0).lexeme() + "\"");
        };
    }

    private Block parseBody() {
        int start = start();

        List<Stmt> stmts = new ArrayList<>();
        if (match(TokenKind.LBRACE)) {
            while (!match(TokenKind.RBRACE)) {
                if (atEnd()) throw exception("Expected '}'");
                stmts.add(parseStmt());
            }
        } else {
            stmts.add(parseStmt());
        }

        return new Block(stmts, range(start));
    }

    private VarDecl parseVarDecl() {
        int start = start();

        if (!match(TokenKind.VAR)) throw new AssertionError();
        String ident = expect(TokenKind.IDENT, "Expected variable identifier").lexeme();
        expect(TokenKind.COLON, "Expected ':' after variable identifier");
        TypeRef typeRef = expectTypeRef();
        expect(TokenKind.ASSIGN, "Expected '=' after variable type");
        Expr init = parseExpr();
        expect(TokenKind.SEMICOLON, "Expected ';' after variable declaration");

        return new VarDecl(ident, typeRef, init, range(start));
    }

    private FuncDef parseFuncDef() {
        int start = start();

        if (!match(TokenKind.FUNC)) throw new AssertionError();
        String ident = expect(TokenKind.IDENT, "Expected function identifier").lexeme();
        expect(TokenKind.LPAREN, "Expected '(' after function identifier");

        List<Param> params = new ArrayList<>();
        if (!check(TokenKind.RPAREN)) {
            do {
                int paramStart = start();

                String paramIdent = expect(TokenKind.IDENT, "Expected parameter identifier").lexeme();
                expect(TokenKind.COLON, "Expected ':' after parameter identifier");
                TypeRef typeRef = expectTypeRef();

                params.add(new Param(paramIdent, typeRef, range(paramStart)));
            } while (match(TokenKind.COMMA));
        }

        expect(TokenKind.RPAREN, "Expected ')' after function parameters");
        expect(TokenKind.COLON, "Expected ':' after function parameters");

        TypeRef returnType = expectTypeRef();
        Block body = parseBody();

        return new FuncDef(ident, params, returnType, body, range(start));
    }

    private Stmt parseStmt() {
        return switch (peek(0).kind()) {
            case VAR -> parseVarDecl();
            case WHILE -> parseWhileStmt();
            case IF -> parseIfStmt();
            case RETURN -> parseReturnStmt();
            case IDENT -> peek(1).kind() == TokenKind.ASSIGN
                    ? parseAssignStmt()
                    : parseExprStmt();

            default -> parseExprStmt();
        };
    }

    private ExprStmt parseExprStmt() {
        int start = start();

        Expr expr = parseExpr();
        expect(TokenKind.SEMICOLON, "Expected ';' after expression");

        return new ExprStmt(expr, range(start));
    }

    private WhileStmt parseWhileStmt() {
        int start = start();

        if (!match(TokenKind.WHILE)) throw new AssertionError();
        expect(TokenKind.LPAREN, "Expected '(' after \"while\"");
        Expr condition = parseExpr();
        expect(TokenKind.RPAREN, "Expected ')' after while condition");
        Block body = parseBody();

        return new WhileStmt(condition, body, range(start));
    }

    private IfStmt parseIfStmt() {
        int start = start();

        if (!match(TokenKind.IF)) throw new AssertionError();
        expect(TokenKind.LPAREN, "Expected '(' after \"if\"");
        Expr condition = parseExpr();
        expect(TokenKind.RPAREN, "Expected ')' after if condition");
        Block thenBranch = parseBody();
        Block elseBranch = match(TokenKind.ELSE) ? parseBody() : null;

        return new IfStmt(condition, thenBranch, elseBranch, range(start));
    }

    private ReturnStmt parseReturnStmt() {
        int start = start();

        if (!match(TokenKind.RETURN)) throw new AssertionError();
        Expr value = check(TokenKind.SEMICOLON) ? null : parseExpr();
        expect(TokenKind.SEMICOLON, "Expected ';' after return statement");

        return new ReturnStmt(value, range(start));
    }

    private AssignStmt parseAssignStmt() {
        int start = start();

        if (!check(TokenKind.IDENT)) throw new AssertionError();
        String ident = advance().lexeme();
        expect(TokenKind.ASSIGN, "Expected '=' after variable identifier");
        Expr value = parseExpr();
        expect(TokenKind.SEMICOLON, "Expected ';' after assign statement");

        return new AssignStmt(ident, value, range(start));
    }

    private Expr parseExpr() {
        return parseExpr(parseOperand(), BinaryOp.MIN_PRECEDENCE);
    }

    private Expr parseExpr(Expr left, int minPrecedence) {
        int start = left.range().start();

        while (BINARY_OPS.containsKey(peek(0).kind())) {
            BinaryOp op = BINARY_OPS.get(peek(0).kind());
            if (op.precedence < minPrecedence) break;
            advance();

            Expr right = parseOperand();
            right = parseExpr(right, op.precedence + 1);

            left = new BinaryExpr(left, op, right, range(start));
        }

        return left;
    }

    private Expr parseOperand() {
        return UNARY_OPS.containsKey(peek(0).kind())
                ? parseUnaryExpr()
                : parsePrimaryExpr();
    }

    private Expr parseUnaryExpr() {
        int start = start();

        UnaryOp op = UNARY_OPS.get(advance().kind());
        if (op == null) throw new AssertionError();
        Expr operand = parseOperand();

        return new UnaryExpr(op, operand, range(start));
    }

    private Expr parsePrimaryExpr() {
        TokenKind kind = peek(0).kind();

        if (kind == TokenKind.IDENT && peek(1).kind() == TokenKind.LPAREN) {
            return parseCallExpr();
        }
        else if (kind == TokenKind.IDENT) {
            return parseVarExpr();
        }
        else if (LITERALS.contains(kind)) {
            return parseLiteral();
        }
        else if (kind == TokenKind.LPAREN) {
            advance();
            Expr expr = parseExpr();
            expect(TokenKind.RPAREN, "Expected ')' after expression");
            return expr;
        }

        throw exception("Unexpected token \"" + peek(0).lexeme() + "\"");
    }

    private CallExpr parseCallExpr() {
        int start = start();

        if (!check(TokenKind.IDENT)) throw new AssertionError();
        String ident = advance().lexeme();
        if (!match(TokenKind.LPAREN)) throw new AssertionError();

        List<Expr> args = new ArrayList<>();
        if (!check(TokenKind.RPAREN)) {
            do {
                args.add(parseExpr());
            } while (match(TokenKind.COMMA));
        }

        expect(TokenKind.RPAREN, "Expected ')' after call arguments");

        return new CallExpr(ident, args, range(start));
    }

    private VarExpr parseVarExpr() {
        int start = start();

        if (!check(TokenKind.IDENT)) throw new AssertionError();
        String ident = advance().lexeme();

        return new VarExpr(ident, range(start));
    }

    private Expr parseLiteral() {
        int start = start();

        Token token = advance();
        TokenKind kind = token.kind();
        if (!LITERALS.contains(kind)) throw new AssertionError();
        String lexeme = token.lexeme();

        return switch (kind) {
            case STR_LIT -> new StrLiteral(lexeme, range(start));
            case INT_LIT -> new IntLiteral(lexeme, range(start));
            case FLOAT_LIT -> new FloatLiteral(lexeme, range(start));
            case TRUE -> new BoolLiteral(true, range(start));
            case FALSE -> new BoolLiteral(false, range(start));

            default -> throw new AssertionError();
        };
    }
}

File: src/parser.rs
use crate::ast::*;
use crate::lexer::{Token, TokenKind};
use crate::span::Spanned;

#[derive(Debug)]
pub(crate) enum Error {
    UnexpectedToken(Token),
    ExpectedToken { expected: TokenKind, found: Token },
}

pub(crate) type Result<T> = std::result::Result<T, Error>;

struct Cursor {
    tokens: Vec<Token>,
    pos: usize,
}

impl Cursor {
    fn new(tokens: Vec<Token>) -> Self {
        assert!(!tokens.is_empty(), "`tokens` must be non-empty");
        assert!(matches!(
            tokens.last().map(|t| t.kind()),
            Some(TokenKind::Eof)
        ));

        Self { tokens, pos: 0 }
    }

    fn is_at_end(&self) -> bool {
        matches!(self.peek().kind(), TokenKind::Eof)
    }

    fn peek(&self) -> &Token {
        self.tokens.get(self.pos).expect("bug: cursor out of bounds")
    }

    fn advance(&mut self) -> &Token {
        let token = &self.tokens[self.pos];

        if !self.is_at_end() {
            self.pos += 1;
        }

        token
    }

    fn check(&self, kind: &TokenKind) -> bool {
        self.peek().kind() == kind
    }

    fn match_kind(&mut self, kind: &TokenKind) -> bool {
        if self.check(kind) {
            self.advance();
            true
        } else {
            false
        }
    }

    fn expect(&mut self, kind: &TokenKind) -> Result<&Token> {
        if self.check(kind) {
            Ok(self.advance())
        } else {
            Err(Error::ExpectedToken {
                expected: kind.clone(),
                found: self.peek().clone()
            })
        }
    }
}

mod precedence {
    use crate::ast::BinaryOp;

    pub(super) const MIN: u8 = 0;

    pub(super) fn of(op: BinaryOp) -> u8 {
        match op {
            BinaryOp::Mul | BinaryOp::Div => 5,
            BinaryOp::Add | BinaryOp::Sub => 4,
            BinaryOp::Lt | BinaryOp::Le | BinaryOp::Gt | BinaryOp::Ge => 3,
            BinaryOp::Eq | BinaryOp::Ne => 2,
            BinaryOp::And => 1,
            BinaryOp::Or => 0
        }
    }
}

pub(crate) struct Parser {
    cursor: Cursor,
}

impl Parser {
    fn peek_kind(&self) -> &TokenKind {
        self.cursor.peek().kind()
    }

    fn parse_name(&mut self) -> Result<Name> {
        let token = self.cursor.advance();

        if let TokenKind::Ident(name) = token.kind() {
            Ok(Name::new(name, token.span()))
        } else {
            Err(Error::ExpectedToken {
                expected: TokenKind::Ident(String::new()),
                found: token.clone()
            })
        }
    }

    fn parse_type(&mut self) -> Result<TypeRef> {
        let name = self.parse_name()?;
        let start = name.span();

        let mut ty = TypeRef::new(
            TypeKind::Name(name),
            start
        );

        // recursively parse ...[]
        while self.cursor.match_kind(&TokenKind::LBrack) {
            let end = self.cursor.expect(&TokenKind::RBrack)?.span();

            let span = ty.span().merge(end);
            ty = TypeRef::new(
                TypeKind::Array(Box::new(ty)),
                span
            );
        }

        Ok(ty)
    }

    fn parse_literal(&mut self) -> Result<Expr> {
        let token = self.cursor.advance();

        let lit = match token.kind() {
            TokenKind::Int(v) => Literal::Int(*v),
            TokenKind::Float(v) => Literal::Float(*v),
            TokenKind::Bool(v) => Literal::Bool(*v),
            TokenKind::Str(v) => Literal::Str(v.clone()),
            _ => return Err(Error::UnexpectedToken(token.clone()))
        };

        Ok(Lit::new(lit, token.span()).into())
    }

    fn peek_binary_op(&self) -> Option<BinaryOp> {
        let op = match self.peek_kind() {
            TokenKind::Star => BinaryOp::Mul,
            TokenKind::Slash => BinaryOp::Div,

            TokenKind::Plus => BinaryOp::Add,
            TokenKind::Minus => BinaryOp::Sub,

            TokenKind::Lt => BinaryOp::Lt,
            TokenKind::Le => BinaryOp::Le,
            TokenKind::Gt => BinaryOp::Gt,
            TokenKind::Ge => BinaryOp::Ge,

            TokenKind::EqEq => BinaryOp::Eq,
            TokenKind::Ne => BinaryOp::Ne,

            TokenKind::And => BinaryOp::And,

            TokenKind::Or => BinaryOp::Or,

            _ => return None
        };

        Some(op)
    }

    fn peek_unary_op(&self) -> Option<UnaryOp> {
        let op = match self.peek_kind() {
            TokenKind::Not => UnaryOp::Not,
            TokenKind::Minus => UnaryOp::Neg,
            _ => return None
        };

        Some(op)
    }
}

impl Parser {
    pub(crate) fn parse(tokens: Vec<Token>) -> Result<Program> {
        Self::new(tokens).parse_program()
    }

    fn new(tokens: Vec<Token>) -> Self {
        Self { cursor: Cursor::new(tokens) }
    }

    fn parse_program(mut self) -> Result<Program> {
        let mut decls = Vec::new();

        while !self.cursor.is_at_end() {
            decls.push(self.parse_decl()?);
        }

        Ok(Program::new(decls))
    }

    fn parse_decl(&mut self) -> Result<Decl> {
        let token = self.cursor.peek();

        match token.kind() {
            TokenKind::Func => self.parse_func().map(Decl::from),
            TokenKind::Var => self.parse_var().map(Decl::from),

            _ => Err(Error::UnexpectedToken(token.clone()))
        }
    }

    fn parse_func(&mut self) -> Result<Func> {
        // func name(params...): return_type { body }

        let start = self.cursor.expect(&TokenKind::Func)?.span();

        let name = self.parse_name()?;
        self.cursor.expect(&TokenKind::LParen)?;

        let mut params = Vec::new();

        if !self.cursor.check(&TokenKind::RParen) {
            loop {
                let name = self.parse_name()?;
                self.cursor.expect(&TokenKind::Colon)?;

                let type_ref = self.parse_type()?;

                let span = name.span().merge(type_ref.span());
                params.push(Param::new(name, type_ref, span));

                if !self.cursor.match_kind(&TokenKind::Comma) {
                    break;
                }
            }
        }

        self.cursor.expect(&TokenKind::RParen)?;
        self.cursor.expect(&TokenKind::Colon)?;

        let return_type = self.parse_type()?;

        let body = self.parse_block()?;
        let span = start.merge(body.span());

        Ok(Func::new(name, params, return_type, body, span))
    }

    fn parse_var(&mut self) -> Result<Var> {
        // var name: type = init;

        let start = self.cursor.expect(&TokenKind::Var)?.span();

        let name = self.parse_name()?;
        self.cursor.expect(&TokenKind::Colon)?;

        let type_ref = self.parse_type()?;
        self.cursor.expect(&TokenKind::Eq)?;

        let init = self.parse_expr()?;
        self.cursor.expect(&TokenKind::Semicolon)?;

        let span = start.merge(init.span());

        Ok(Var::new(name, type_ref, init, span))
    }

    fn parse_stmt(&mut self) -> Result<Stmt> {
        let token = self.cursor.peek();

        match token.kind() {
            TokenKind::If => self.parse_if().map(Stmt::from),
            TokenKind::While => self.parse_while().map(Stmt::from),
            TokenKind::Return => self.parse_return().map(Stmt::from),

            TokenKind::Var => self.parse_var()
                .map(Decl::from)
                .map(|decl| {
                    let span = decl.span();
                    DeclStmt::new(decl, span)
                })
                .map(Stmt::from),

            _ => self.parse_assign_or_expr()
        }
    }

    fn parse_block(&mut self) -> Result<Block> {
        // allows single stmt block without braces
        // like `if cond stmt`

        let start = self.cursor.peek().span();
        let mut stmts = Vec::new();

        if self.cursor.match_kind(&TokenKind::LBrace) {
            while !self.cursor.check(&TokenKind::RBrace) {
                stmts.push(self.parse_stmt()?);
            }

            let end = self.cursor.expect(&TokenKind::RBrace)?.span();
            Ok(Block::new(stmts, start.merge(end)))
        } else {
            let stmt = self.parse_stmt()?;
            let span = start.merge(stmt.span());

            stmts.push(stmt);
            Ok(Block::new(stmts, span))
        }
    }

    fn parse_if(&mut self) -> Result<If> {
        let start = self.cursor.expect(&TokenKind::If)?.span();

        let condition = self.parse_expr()?;
        let then_body = self.parse_block()?;

        let (else_body, end) =
            if self.cursor.match_kind(&TokenKind::Else) {

            let block = self.parse_block()?;
            let end = block.span();
            (Some(block), end)
        } else {
            (None, then_body.span())
        };

        Ok(If::new(condition, then_body, else_body, start.merge(end)))
    }

    fn parse_while(&mut self) -> Result<While> {
        let start = self.cursor.expect(&TokenKind::While)?.span();

        let condition = self.parse_expr()?;
        let body = self.parse_block()?;

        let span = start.merge(body.span());

        Ok(While::new(condition, body, span))
    }

    fn parse_return(&mut self) -> Result<Return> {
        let start = self.cursor.expect(&TokenKind::Return)?.span();

        let value = if self.cursor.check(&TokenKind::Semicolon) {
            None
        } else {
            Some(self.parse_expr()?)
        };

        let end = self.cursor.expect(&TokenKind::Semicolon)?.span();

        Ok(Return::new(value, start.merge(end)))
    }

    fn parse_assign_or_expr(&mut self) -> Result<Stmt> {
        let expr = self.parse_expr()?;
        let span = expr.span();

        if self.cursor.match_kind(&TokenKind::Eq) {
            let value = self.parse_expr()?;
            let span = span.merge(value.span());

            self.cursor.expect(&TokenKind::Semicolon)?;
            Ok(Assign::new(expr, value, span).into())
        } else {
            self.cursor.expect(&TokenKind::Semicolon)?;
            Ok(ExprStmt::new(expr, span).into())
        }
    }

    fn parse_expr(&mut self) -> Result<Expr> {
        let operand = self.parse_operand()?;
        self.parse_expr_inner(operand, precedence::MIN)
    }

    fn parse_expr_inner(&mut self, mut lhs: Expr, min_precedence: u8) -> Result<Expr> {
        // continue consuming binary ops as long as they bind tighter
        // than the current `min_precedence`
        while let Some(op) = self.peek_binary_op() {
            let precedence = precedence::of(op);

            // stop if this op would bind too weak for this expr
            if precedence <= min_precedence {
                break;
            } else {
                self.cursor.advance();
            }

            // parse the rhs operand and recursively parse to allow
            // higher precedence ops to bind
            // note: this assumes all ops are left associative
            let rhs = self.parse_operand()?;
            let rhs = self.parse_expr_inner(rhs, precedence)?;

            // now combine the two into a binary expr
            let span = lhs.span().merge(rhs.span());
            let binary = Binary::new(
                Box::new(lhs),
                op,
                Box::new(rhs),
                span
            );

            lhs = Expr::Binary(binary);
        }

        Ok(lhs)
    }

    fn parse_operand(&mut self) -> Result<Expr> {
        let mut expr = if let Some(op) = self.peek_unary_op() {
            let start = self.cursor.advance().span();

            let expr = self.parse_operand()?;
            let span = start.merge(expr.span());

            Unary::new(op, Box::new(expr), span).into()
        } else {
            self.parse_primary()?
        };

        expr = self.parse_postfix(expr)?;
        Ok(expr)
    }

    fn parse_primary(&mut self) -> Result<Expr> {
        match self.peek_kind() {
            TokenKind::LParen => self.parse_group(),

            TokenKind::Ident(_) => self.parse_ident().map(Expr::from),

            // note: this creates 2 sources of truth for what is a literal,
            // here and in `parse_literal`, but it's a tradeoff that
            // can't be avoided easily
            TokenKind::Int(_) |
            TokenKind::Float(_) |
            TokenKind::Str(_) |
            TokenKind::Bool(_) => self.parse_literal(),

            _ => Err(Error::UnexpectedToken(
                self.cursor.peek().clone()))
        }
    }

    fn parse_group(&mut self) -> Result<Expr> {
        self.cursor.expect(&TokenKind::LParen)?;
        let expr = self.parse_expr()?;
        self.cursor.expect(&TokenKind::RParen)?;
        Ok(expr)
    }

    fn parse_ident(&mut self) -> Result<Ident> {
        let name = self.parse_name()?;
        let span = name.span();
        Ok(Ident::new(name, span))
    }

    fn parse_postfix(&mut self, mut expr: Expr) -> Result<Expr> {
        loop {
            expr = match self.peek_kind() {
                TokenKind::LParen => self.parse_call(expr)?.into(),
                TokenKind::LBrack => self.parse_index(expr)?.into(),
                TokenKind::Dot => self.parse_member(expr)?.into(),
                _ => break
            }
        }

        Ok(expr)
    }

    fn parse_call(&mut self, expr: Expr) -> Result<Call> {
        self.cursor.expect(&TokenKind::LParen)?;
        let mut args = Vec::new();

        if !self.cursor.check(&TokenKind::RParen) {
            loop {
                let arg = self.parse_expr()?;
                args.push(arg);

                if !self.cursor.match_kind(&TokenKind::Comma) {
                    break;
                }
            }
        }

        let end = self.cursor.expect(&TokenKind::RParen)?.span();
        let span = expr.span().merge(end);

        Ok(Call::new(
            Box::new(expr),
            args,
            span
        ))
    }

    fn parse_index(&mut self, expr: Expr) -> Result<Index> {
        self.cursor.expect(&TokenKind::LBrack)?;

        let index = self.parse_expr()?;

        let end = self.cursor.expect(&TokenKind::RBrack)?.span();
        let span = expr.span().merge(end);

        Ok(Index::new(
            Box::new(expr),
            Box::new(index),
            span
        ))
    }

    fn parse_member(&mut self, expr: Expr) -> Result<Member> {
        self.cursor.expect(&TokenKind::Dot)?;

        let member = self.parse_name()?;
        let span = expr.span().merge(member.span());

        Ok(Member::new(
            Box::new(expr),
            member,
            span
        ))
    }
}
